crates/pathfinder/src/rpc/v02/method/get_proof.rs

```diff
@@ -0,0 +1,272 @@
+use anyhow::{anyhow, Context};
+use serde::Deserialize;
+
+use crate::core::{BlockId, ContractAddress, StorageAddress};
+use crate::rpc::v02::RpcContext;
+use crate::state::merkle_tree::ProofNode;
+use crate::state::state_tree::{ContractsStateTree, GlobalStateTree};
+use crate::storage::{ContractsStateTable, StarknetBlocksBlockId, StarknetBlocksTable};
+
+#[derive(Deserialize, Debug, PartialEq, Eq)]
+pub struct GetStorageProofInput {
+    pub contract_address: ContractAddress,
+    pub key: StorageAddress,
+    pub block_id: BlockId,
+}
+
+crate::rpc::error::generate_rpc_error_subset!(
+    GetStorageProofError: ContractNotFound,
+    BlockNotFound
+);
+
+/// TODO: description
+pub async fn get_proof(
+    context: RpcContext,
+    input: GetStorageProofInput,
+) -> Result<Vec<ProofNode>, GetStorageProofError> {
```

Comment:
For the output just `Vec<ProofNode>` will be insufficient. We'll need something like:
```rust
pub struct GetProofOutput {
  // This is the global root -> contract state hash proof
  contract_proof: Vec<ProofNode>,

  // Need to include the other meta-data that lets the caller verify the
  // contract state hash to contract root calculation.
  nonce: ContractNonce,
  // Technically they can get this from `storage_proofs`, but that can be annoying for the caller.
  root: ContractRoot,
  // This is currently just a constant = 0. Unclear if this should be included at all.
  // We can probably just skip this for now. Will check with StarkWare.
  contract_state_hash_version: StarkHash,

  // Assuming we allow multiple queries in one. Contract root -> storage key proofs
  storage_proofs: Vec<Vec<ProofNode>>,
}
```

Reply:
Checked with StarkWare and they suggest we do add in `contract_state_hash_version` so that the user can just do chain hashing `H(H(H(hash, root), nonce), CONTRACT_STATE_HASH_VERSION)` without worrying about random constants that can change in the future.

Reply:
Agreed for the output. We mentioned it in the issue. I didn't do it here because I thought the RPC work was "put on hold". Will work on it now (and update the issue :p).

Reply:
By the way @Mirko-von-Leipzig I'm thinking about moving to something like:
```rust
pub struct GetStorageProofOutput {
    // This is the global root -> contract state hash proof
    contract_proof: Vec<ProofNode>,

    // Required by the verifier to verify the contract state hash to contract root calculation.
    class_hash: Option<ClassHash>,
    nonce: Option<ContractNonce>,

    // Root of the Contract state tree (if it exists).
    root: Option<ContractRoot>,

    contract_state_hash_version: Option<StarkHash>,

    // Assuming we allow multiple queries in one. Contract root -> storage key proofs
    storage_proofs: Option<Vec<Vec<ProofNode>>>,
}
```

Meaning putting `Option`s everywhere, in case the contract state tree isn't found. Would that make sense to you?

Reply:
Hmm. Yeah makes sense. Although I think then bundling it all into a single Option makes more semantic sense. So another struct with all the optionals bundled inside.

Reply:
see 49163a93bcf700b127e25241e5b0e2bd0f3eba1a